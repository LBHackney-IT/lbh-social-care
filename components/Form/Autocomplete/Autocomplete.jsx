import Downshift from 'downshift';
import PropTypes from 'prop-types';
import cx from 'classnames';
import { Controller } from 'react-hook-form';

import { sortObject } from 'utils/objects';
import ErrorMessage from 'components/ErrorMessage/ErrorMessage';
import ClearIcon from 'components/Icons/Clear';
import style from './Autocomplete.module.scss';

const AutocompleteComponents = ({
  label,
  hint,
  name,
  error,
  labelSize = 'm',
  required,
  onChange,
  placeholder,
  width = 20,
  govGrid,
  inputClassName,
  options,
}) => {
  const optionObject = options.map((option) => {
    if (typeof option === 'string') {
      return { value: option, text: option };
    }
    return option;
  });

  const items = sortObject(optionObject);
  return (
    <Downshift
      onChange={(selection) => onChange(selection ? selection.value : '')}
      itemToString={(item) => (item ? item.text : '')}
    >
      {({
        getInputProps,
        getItemProps,
        getMenuProps,
        isOpen,
        inputValue,
        getRootProps,
        toggleMenu,
        highlightedIndex,
        selectedItem,
        clearSelection,
      }) => (
        <div
          className={cx(`govuk-form-group govuk-grid-column-${govGrid}`, {
            'govuk-form-group--error': error,
          })}
        >
          <label
            className={`govuk-label govuk-label--${labelSize}`}
            htmlFor={name}
          >
            {label} {required && <span className="govuk-required">*</span>}
          </label>

          <div
            className={style.inputContainer}
            {...getRootProps({}, { suppressRefError: true })}
          >
            {hint && (
              <span id={`${name}-hint`} className="govuk-hint">
                {hint}
              </span>
            )}
            {error && <ErrorMessage label={error.message} />}
            <input
              {...getInputProps()}
              placeholder={placeholder || 'Select or search for a value'}
              id={name}
              data-testid={name}
              name={`${name}.autocomplete`}
              onClick={toggleMenu}
              aria-describedby={hint && `${name}-hint`}
              className={cx(`govuk-input`, inputClassName, {
                [`govuk-input--width-${width}`]: width,
              })}
              style={
                error ? { borderColor: '#d4351c' } : { borderColor: '#000' }
              }
            />
            {inputValue && (
              <span onClick={clearSelection} className={style.clear}>
                <ClearIcon />
              </span>
            )}
          </div>
          <ul
            {...getMenuProps()}
            className={style.suggestionList}
            style={
              isOpen
                ? { borderBottom: '1px solid #aaa' }
                : { borderBottom: 'none' }
            }
          >
            {isOpen &&
              items
                .filter(
                  (item) =>
                    !inputValue ||
                    item.text.toLowerCase().includes(inputValue.toLowerCase())
                )
                .map((item, index) => (
                  <li
                    key={index}
                    data-testid={`${name}_${index}`}
                    className={style.suggestion}
                    {...getItemProps({
                      key: index,
                      index,
                      item,
                      style: {
                        color: highlightedIndex === index ? '#fff' : '#000',
                        backgroundColor:
                          highlightedIndex === index ? '#8b8b8b' : '#fff',
                        fontWeight: selectedItem === item ? 'bold' : 'normal',
                      },
                    })}
                  >
                    {item.text}
                  </li>
                ))}
          </ul>
        </div>
      )}
    </Downshift>
  );
};

AutocompleteComponents.propTypes = {
  name: PropTypes.string.isRequired,
  label: PropTypes.string.isRequired,
  labelSize: PropTypes.string,
  hint: PropTypes.string,
  onChange: PropTypes.func,
  placeholder: PropTypes.string,
  options: PropTypes.array,
  rules: PropTypes.shape({}),
  width: PropTypes.number,
  inputClassName: PropTypes.string,
  govGrid: PropTypes.string,
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  required: PropTypes.oneOfType([PropTypes.bool, PropTypes.string]),
  error: PropTypes.shape({
    message: PropTypes.string.isRequired,
  }),
};

const Autocomplete = ({ name, control, ...otherProps }) => (
  <Controller
    render={({ onChange, value, name }) => (
      <AutocompleteComponents
        name={name}
        value={value}
        onChange={(value) => onChange(value)}
        {...otherProps}
      />
    )}
    control={control}
    name={name}
    defaultValue={control?.defaultValuesRef?.current[name] || ''}
  />
);

Autocomplete.propTypes = {
  name: PropTypes.string.isRequired,
  control: PropTypes.object.isRequired,
  onChange: PropTypes.func,
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
};
export default Autocomplete;
