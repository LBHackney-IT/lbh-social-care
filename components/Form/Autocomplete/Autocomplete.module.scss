@import 'lbh-frontend/lbh/base';

.inputContainer {
  position: relative;
  display: flex;
  max-width: 48ex;
  margin-top: 0;
}

.toggle,
.clear {
  cursor: pointer;
  position: relative;
  z-index: 4;
  right: 32px;
  padding: 4px 8px;
}

.suggestionList {
  box-sizing: border-box;
  background: lbh-colour('lbh-white');
  border: 2px solid lbh-colour('lbh-text');
  position: absolute;
  top: 14px;
  left: 0px;
  width: 100%;
  list-style: none;
  z-index: 2;
  max-height: 300px;
  overflow-y: auto;
}

.suggestion {
  margin-top: 0;
  @include lbh-body-m;
  padding: 5px 10px;
  line-height: 1.2;

  &[aria-selected='true'] {
    background: lbh-colour('lbh-link');
    color: lbh-colour('lbh-white');
  }
}

.button {
  cursor: pointer;
  margin-top: 0;
  position: absolute;
  right: 2px;
  top: 2px;
  bottom: 2px;
  display: flex;
  justify-content: center;
  align-items: center;
  background: none;
  border: none;
  width: 35px;
  &:focus {
    outline: none;
    background: lbh-colour('lbh-primary-focus');
  }
}

.list {
  box-sizing: border-box;
  background: lbh-colour('lbh-white');
  border: 2px solid lbh-colour('lbh-text');
  position: absolute;
  top: 14px;
  left: 0px;
  width: 100%;
  list-style: none;
  z-index: 2;
  max-height: 300px;
  overflow-y: auto;

  &:empty {
    border: none;
  }
}

.option {
  margin-top: 0;
  @include lbh-body-m;
  padding: 5px 10px;
  line-height: 1.2;

  &[aria-selected='true'] {
    background: lbh-colour('lbh-link');
    color: lbh-colour('lbh-white');
  }
}

.noResults {
  padding: 5px 10px;
  @include lbh-body-m;
  color: lbh-colour('lbh-secondary-text');
}
