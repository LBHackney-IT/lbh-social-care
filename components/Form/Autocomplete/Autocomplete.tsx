import { forwardRef, useEffect, useRef, useMemo } from 'react';
import Downshift from 'downshift';
import cx from 'classnames';
import { Controller, Control } from 'react-hook-form';

import { sortObject } from 'utils/objects';
import ErrorMessage from 'components/ErrorMessage/ErrorMessage';
import style from './Autocomplete.module.scss';

import {
  Autocomplete as IAutocomplete,
  ObjectOption,
  Option,
} from 'components/Form/types';
interface AutoProps extends Omit<IAutocomplete, 'control'> {
  value?: string | number | null;
  text?: string;
  inputClassName?: string;
  onChange: (arg0: string | number | null) => string | number | null | void;
}
interface Props extends IAutocomplete {
  control: Control;
}

const convertToObject = (options: Option[]) => {
  const objects = options?.map((option) => {
    if (typeof option === 'string') {
      return { value: option, text: option };
    }
    return option;
  });
  return sortObject(objects);
};

const findValue = (items: ObjectOption[], value: string) =>
  items.find((item) => String(item.value) === value);

// eslint-disable-next-line react/display-name
export const Autocomplete = forwardRef<HTMLInputElement, AutoProps>(
  (
    {
      label,
      hint,
      name,
      error,
      labelSize = 'm',
      required,
      onChange,
      placeholder,
      width = 20,
      govGrid,
      value,
      inputClassName,
      options,
      ...otherProps
    }: AutoProps,
    ref
  ) => {
    const items = useMemo(() => convertToObject(options), [options]);
    const autoRef = useRef<any>(null);
    const getTextValue = () => {
      if (value) {
        const selection = findValue(items, String(value));
        return selection?.text;
      }
      return '';
    };
    useEffect(() => {
      if (value === null) {
        autoRef.current?.clearSelection();
      }
    }, [value]);
    useEffect(() => {
      value &&
        !findValue(items, String(value)) &&
        autoRef.current?.clearSelection();
      // we only need to check {items}
      // eslint-disable-next-line react-hooks/exhaustive-deps
    }, [items]);
    return (
      <Downshift
        initialInputValue={getTextValue()}
        onChange={(selection) =>
          onChange && onChange(selection ? selection.value : '')
        }
        itemToString={(item) => (item ? item.text : '')}
        ref={autoRef}
      >
        {({
          getInputProps,
          getItemProps,
          getMenuProps,
          isOpen,
          inputValue,
          toggleMenu,
          highlightedIndex,
          getToggleButtonProps,
          clearSelection,
        }) => {
          const inputProps = getInputProps();
          return (
            <div
              className={cx(`lbh-form-group govuk-form-group`, {
                [`govuk-grid-column-${govGrid}`]: govGrid,
                'govuk-form-group--error': error,
              })}
            >
              <label
                className={`govuk-label lbh-label`}
                id={inputProps['aria-labelledby']}
              >
                {label} {required && <span className="lbh-required">*</span>}
              </label>
              <div className={style.inputContainer}>
                {hint && (
                  <span id={`${name}-hint`} className="govuk-hint lbh-hint">
                    {hint}
                  </span>
                )}
                {error && <ErrorMessage label={error.message} />}
                <input
                  {...inputProps}
                  id={name}
                  placeholder={placeholder || 'Select or search for name'}
                  data-testid={name}
                  onClick={(): void => {
                    toggleMenu();
                  }}
                  aria-describedby={hint && `${name}-hint`}
                  className={cx(`lbh-input govuk-input`, inputClassName, {
                    [`govuk-input--width-${width} `]: width,
                  })}
                  ref={ref}
                  {...otherProps}
                />
                <button
                  {...getToggleButtonProps()}
                  aria-label="toggle menu"
                  className={style.button}
                >
                  <svg width="17" height="10" viewBox="0 0 17 10" fill="none">
                    <path
                      d="M2 1.5L8.5 7.5L15 1.5"
                      stroke="#0B0C0C"
                      strokeWidth="3"
                    />
                  </svg>
                </button>

                <ul
                  {...getMenuProps()}
                  className={style.list}
                  style={
                    isOpen
                      ? { borderBottom: '1px solid #aaa' }
                      : { borderBottom: 'none' }
                  }
                >
                  {isOpen &&
                    items
                      .filter(
                        (item) =>
                          !inputValue ||
                          item.text
                            .toLowerCase()
                            .includes(inputValue.toLowerCase())
                      )
                      .map((item, index) => (
                        <li
                          key={index}
                          data-testid={`${name}_${index}`}
                          className={style.option}
                          {...getItemProps({
                            key: `${item}-${index}`,
                            index,
                            item,
                            style: {
                              color:
                                highlightedIndex === index ? '#fff' : '#000',
                              backgroundColor:
                                highlightedIndex === index ? '#8b8b8b' : '#fff',
                            },
                          })}
                        >
                          {item.text}
                        </li>
                      ))}
                </ul>
              </div>
            </div>
          );
        }}
      </Downshift>
    );
  }
);

const ControlledAutocomplete = ({
  name,
  control,
  rules,
  ...otherProps
}: Props): React.ReactElement => (
  <Controller
    render={({ onChange, value }) => (
      <Autocomplete
        name={name}
        value={value}
        onChange={onChange}
        {...otherProps}
      />
    )}
    name={name}
    rules={{
      ...rules,
      validate: {
        ...(typeof rules?.validate === 'function'
          ? { validation: rules.validate }
          : rules?.validate),
      },
    }}
    control={control}
    defaultValue={control.defaultValuesRef.current[name] || ''}
  />
);

export default ControlledAutocomplete;
function getToggleButtonProps(): JSX.IntrinsicAttributes &
  import('react').ClassAttributes<HTMLButtonElement> &
  import('react').ButtonHTMLAttributes<HTMLButtonElement> {
  throw new Error('Function not implemented.');
}
