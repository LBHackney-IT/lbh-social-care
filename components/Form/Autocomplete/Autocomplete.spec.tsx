import { fireEvent, render } from '@testing-library/react';

import { Autocomplete } from './Autocomplete';

describe(`Autocomplete`, () => {
  const props = {
    label: 'Foo',
    name: 'autocomplete',
    onChange: jest.fn(),
    placeholder: 'Foo bar',
    options: [
      { text: 'foo', value: 14 },
      { text: 'bar', value: '10' },
    ],
  };

  it('should render properly', () => {
    const { getByTestId, asFragment } = render(<Autocomplete {...props} />);
    const autocomplete = getByTestId('autocomplete');

    fireEvent.click(autocomplete);
    expect(autocomplete).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it('should select value', () => {
    const { getByTestId, getAllByRole } = render(<Autocomplete {...props} />);
    const autocomplete = getByTestId('autocomplete');

    fireEvent.click(autocomplete);
    expect(getAllByRole('option').length).toBe(2);
    fireEvent.change(autocomplete, { target: { value: 'f' } });
    expect(getAllByRole('option').length).toBe(1);
    fireEvent.click(getByTestId('autocomplete_0'));
    expect(autocomplete).toHaveValue('foo');
  });
});
