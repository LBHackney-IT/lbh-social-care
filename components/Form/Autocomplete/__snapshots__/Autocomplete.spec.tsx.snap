// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Autocomplete should render properly 1`] = `
<DocumentFragment>
  <div
    aria-expanded="true"
    aria-haspopup="listbox"
    aria-labelledby="downshift-0-label"
    aria-owns="downshift-0-menu"
    class="lbh-form-group govuk-form-group"
    role="combobox"
  >
    <label
      class="govuk-label lbh-label"
      id="downshift-0-label"
    >
      Foo 
    </label>
    <div
      class="inputContainer"
    >
      <input
        aria-autocomplete="list"
        aria-controls="downshift-0-menu"
        aria-labelledby="downshift-0-label"
        autocomplete="off"
        class="lbh-input govuk-input govuk-input--width-20 "
        data-testid="autocomplete"
        id="autocomplete"
        placeholder="Foo bar"
        value=""
      />
      <button
        aria-haspopup="true"
        aria-label="toggle menu"
        class="button"
        data-toggle="true"
        role="button"
        type="button"
      >
        <svg
          fill="none"
          height="10"
          viewBox="0 0 17 10"
          width="17"
        >
          <path
            d="M2 1.5L8.5 7.5L15 1.5"
            stroke="#0B0C0C"
            stroke-width="3"
          />
        </svg>
      </button>
      <ul
        aria-labelledby="downshift-0-label"
        class="list"
        id="downshift-0-menu"
        role="listbox"
        style="border-bottom: 1px solid #aaa;"
      >
        <li
          aria-selected="false"
          class="option"
          data-testid="autocomplete_0"
          id="downshift-0-item-0"
          role="option"
          style="color: rgb(0, 0, 0); background-color: rgb(255, 255, 255);"
        >
          bar
        </li>
        <li
          aria-selected="false"
          class="option"
          data-testid="autocomplete_1"
          id="downshift-0-item-1"
          role="option"
          style="color: rgb(0, 0, 0); background-color: rgb(255, 255, 255);"
        >
          foo
        </li>
      </ul>
    </div>
  </div>
</DocumentFragment>
`;
